crates/bevy_render/src/batching/gpu_preprocessing.rs

```diff
@@ -0,0 +1,313 @@
+    pub current_input_buffer: BufferVec<BDI>,
+
+    /// The uniform data inputs for the previous frame.
+    ///
+    /// The indices don't generally line up between `current_input_buffer`
+    /// and `previous_input_buffer`, because, among other reasons, entities
+    /// can spawn or despawn between frames. Instead, each current buffer
+    /// data input uniform is expected to contain the index of the
+    /// corresponding buffer data input uniform in this list.
+    pub previous_input_buffer: BufferVec<BDI>,
+}
+
+/// One invocation of the preprocessing shader: i.e. one mesh instance in a
+/// view.
+#[derive(Clone, Copy, Pod, Zeroable, ShaderType)]
+#[repr(C)]
+pub struct PreprocessWorkItem {
+    /// The index of the batch input data in the input buffer that the shader
+    /// reads from.
+    pub input_index: u32,
+    /// The index of the `MeshUniform` in the output buffer that we write to.
+    pub output_index: u32,
+}
+
+impl<BD, BDI> BatchedInstanceBuffers<BD, BDI>
+where
+    BD: GpuArrayBufferable + Sync + Send + 'static,
+    BDI: Pod,
+{
+    /// Creates new buffers.
+    pub fn new() -> Self {
+        BatchedInstanceBuffers {
+            data_buffer: UninitBufferVec::new(BufferUsages::STORAGE),
+            work_item_buffers: EntityHashMap::default(),
+            current_input_buffer: BufferVec::new(BufferUsages::STORAGE),
+            previous_input_buffer: BufferVec::new(BufferUsages::STORAGE),
+        }
+    }
+
+    /// Returns the binding of the buffer that contains the per-instance data.
+    ///
+    /// This buffer needs to be filled in via a compute shader.
+    pub fn instance_data_binding(&self) -> Option<BindingResource> {
+        self.data_buffer
+            .buffer()
+            .map(|buffer| buffer.as_entire_binding())
+    }
+}
+
+impl<BD, BDI> Default for BatchedInstanceBuffers<BD, BDI>
+where
+    BD: GpuArrayBufferable + Sync + Send + 'static,
+    BDI: Pod,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A system that removes GPU preprocessing work item buffers that correspond to
+/// deleted [`ViewTarget`]s.
+///
+/// This is a separate system from [`super::clear_batched_instance_buffers`]
+/// because [`ViewTarget`]s aren't created until after the extraction phase is
+/// completed.
+pub fn delete_old_work_item_buffers<GFBD>(
+    mut gpu_batched_instance_buffers: ResMut<
+        BatchedInstanceBuffers<GFBD::BufferData, GFBD::BufferInputData>,
+    >,
+    view_targets: Query<Entity, With<ViewTarget>>,
+) where
+    GFBD: GetFullBatchData,
+{
+    gpu_batched_instance_buffers
+        .work_item_buffers
+        .retain(|entity, _| view_targets.contains(*entity));
+}
+
+/// Batch the items in a sorted render phase, when GPU instance buffer building
+/// isn't in use. This means comparing metadata needed to draw each phase item
+/// and trying to combine the draws into a batch.
+pub fn batch_and_prepare_sorted_render_phase<I, GFBD>(
```

Comment:
I wonder if there is any way to make more of this code be shared between the CPU and GPU paths, for the sake of maintainability and not having to remember to update two implementations.

Reply:
I factored out most of it.